### PR TITLE
Swift: first version of query targeting weak hashing

### PR DESCRIPTION
**Pull request by redsun82**
This is a first version of the weak hashing query targeting CWE-328.

It only considers use of `CryptoKit` (and the `Crypto` open source implementation) for now.

### CODE REVIEW

**Comment by redsun82** [comment]
As can be deduced from [a5233c0](https://github.com/github/codeql/pull/10335/commits/a5233c0d01c3e9c18a987f5994b69f8eb607edd1), the tags and the `qhelp` files are heavily influenced by (= partially copied from) the `py/weak-sensitive-data-hashing` query

File: swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.ql
Comment (geoffw0) [inline swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.ql:2]: I keep wondering if we can widen the target 'on sensitive data' somehow, because I think we might miss quite a lot of good results due to this constraint.  I understand that its OK to use insecure hashes (such as MD5) for non-security purposes though, and we don't want to report those.  Also the Python version of this query seems to involve sensitive data somehow.

File: swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.qhelp
Comment (geoffw0) [inline swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.qhelp:52]: It would be good to clarify the phrase "limited input space" somewhere in this doc.

**Review by geoffw0 — COMMENTED** [review]
This looks great.  I've made a few comments but nothing major.

Are there other libraries than `CryptoKit` we will want to add support for (preferably after getting what we have here merged)?

File: swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.ql
Comment (redsun82) [reply swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.ql:2]: Yeah, I think for example MD5 hashing is still somewhat commonplace to have a rough comparison check on files, so there might be some uses out there.

File: swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift
Comment (redsun82) [reply swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:68]: although, I would probably not yet include passwords hashed by SHA-2 as apparently that should also be flagged

Comment (geoffw0) [reply swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:68]: Why not both SHA-2 and something definitely good?  Its OK to have test cases we don't yet identify (and mark them `[NOT DETECTED]`).

Comment (redsun82) [reply swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:68]: good point. Although, as far as I can tell CryptoKit does not give anything stronger

**Comment by redsun82** [comment]
I'm making this ready for review as I'd rather extend the library support in another PR

About the `NOT DETECTED` parts, that should be solved after we use the shared sensitive data heuristics I think.

File: swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift
Comment (geoffw0) [reply swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:146]: If we're sure its broken, I think we should report it even if CryptoKit offers nothing better.  But this is a little odd - we should probably check we're sure its broken first.

File: swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift
Comment (geoffw0) [inline swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:75]: This line looks like it should use `SHA384`?

File: swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift
Comment (geoffw0) [inline swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:81]: ... and this one `SHA512`?

File: swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift
Comment (geoffw0) [inline swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:73]: Are these ones considered `GOOD` because passwords require a stronger algorithm, but `SHA256` is OK for other data?  If this is the case, please could we have a more detailed comment to track that reason.

File: swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift
Comment (geoffw0) [inline swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:166]: What *is* good hashing for a password?

**Review by geoffw0 — COMMENTED** [review]
Happy with the changes to `SensitiveExpr.qll`. :+1:

File: swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift
Comment (redsun82) [reply swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:75]: whoops

File: swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift
Comment (redsun82) [reply swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:166]: any hash that is somewhat hard to compute, in a way that is ok for normal usage but precludes computing billions of hashes in a reasonable time. There are more details on this OWASP cheat sheet page (that is linked in the qhelp file :slightly_smiling_face:):

https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html

File: swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift
Comment (redsun82) [reply swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:81]: double whoops!

File: swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift
Comment (redsun82) [reply swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:73]: will do on Monday :+1:

File: swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift
Comment (redsun82) [reply swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift:166]: that's something that I did not know and actually surprised me, and that I learned from the python/ruby/go version of the query

Comment (geoffw0) [reply swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift]: Thanks, I see what's going on now.

I guess it isn't an issue right now, because we only flag MD5 and SHA1 - but if at some point we flag password hashing using stronger algorithms, I think we'll want to have a test and an example in the qhelp showing exactly how to do it correctly.

To be clear, this is not a change request for this PR.

**Review by geoffw0 — DISMISSED** [review]
I'm very happy with this PR / query.

We should wait for the docs review before merging.

File: swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.qhelp
Comment (hubwriter) [inline swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.qhelp:20]: I'd suggest altering the formatting slightly:
```suggestion
                <b>Pre-image attacks</b>. If you know a hash value <code>h(x)</code>,
                you should not be able to easily find the input <code>x</code>.
            </li>
            <li>
                <b>Collision attacks</b>. If you know a hash value <code>h(x)</code>,
```

File: swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.qhelp
Comment (hubwriter) [inline swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.qhelp:56]: I'd suggest changing this slightly:
```suggestion
            Ensure that you use a strong, modern cryptographic hash function, such as:
        </p>

        <ul>
            <li>
                Argon2, scrypt, bcrypt, or PBKDF2 for passwords and other data with limited input space where
                a dictionary-like attack is feasible.
            </li>
            <li>
                SHA-2, or SHA-3 in other cases.
```

File: swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.qhelp
Comment (hubwriter) [inline swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.qhelp:72]:
```suggestion
            Here is the same function using SHA-512, which is a strong cryptographic hashing function.
```

**Review by hubwriter — DISMISSED** [review]
Docs team review.

Looks good. 👍 
Just a few small comments.

File: swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.qhelp
Comment (redsun82) [reply swift/ql/src/queries/Security/CWE-328/WeakSensitiveDataHashing.qhelp:31]: ok!
```suggestion
```

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Pull request merged**
